Allow declaration of tailscale api endpoint

kelcya opened the pull request:
**What this PR does / why we need it**:
Allow users to declare the baseurl in the provider block. Add a new baseurl variable.

**Which issue this PR fixes** *(use `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 
fixes https://github.com/davidsbond/terraform-provider-tailscale/issues/97

**Special notes for your reviewer**:

davidsbond commented on tailscale/provider.go:
```diff
@@ -31,6 +31,11 @@ func Provider(options ...ProviderOption) *schema.Provider {
 				DefaultFunc: schema.EnvDefaultFunc("TAILSCALE_TAILNET", nil),
 				Required:    true,
 			},
+			"baseurl": {
```
```suggestion
			"base_url": {
```

davidsbond commented on tailscale/provider.go:
```diff
@@ -31,6 +31,11 @@ func Provider(options ...ProviderOption) *schema.Provider {
+			"baseurl": {
+				Type:        schema.TypeString,
+				DefaultFunc: schema.EnvDefaultFunc("TAILSCALE_BASEURL", nil),
+				Required:    true,
```
I think this should be optional, and the default use `https://api.tailscale.com`

davidsbond commented on tailscale/provider.go:
```diff
@@ -31,6 +31,11 @@ func Provider(options ...ProviderOption) *schema.Provider {
 			},
+			"baseurl": {
+				Type:        schema.TypeString,
+				DefaultFunc: schema.EnvDefaultFunc("TAILSCALE_BASEURL", nil),
```
```suggestion
				DefaultFunc: schema.EnvDefaultFunc("TAILSCALE_BASE_URL", nil),
```

davidsbond commented on tailscale/provider.go:
```diff
@@ -59,8 +64,9 @@ func Provider(options ...ProviderOption) *schema.Provider {
 	tailnet := d.Get("tailnet").(string)
-
-	client, err := tailscale.NewClient(apiKey, tailnet)
+	baseurl := d.Get("baseurl").(string)
```
```suggestion
	baseurl := d.Get("base_url").(string)
```

davidsbond commented on tailscale/provider.go:
```diff
@@ -59,8 +64,9 @@ func Provider(options ...ProviderOption) *schema.Provider {
-
-	client, err := tailscale.NewClient(apiKey, tailnet)
+	baseurl := d.Get("baseurl").(string)
+	newbase := tailscale.WithBaseURL(baseurl)
+	client, err := tailscale.NewClient(apiKey, tailnet, newbase)
```
```suggestion
	client, err := tailscale.NewClient(apiKey, tailnet, 
            tailscale.WithBaseURL(baseurl),
	)
```

davidsbond reviewed (commented):
Thanks for raising this, I've suggested a few changes.

I would also be good if you could update the documentation regarding the provider configuration values.

kelcya commented:
@davidsbond I added your suggestions, updated documentation and added the default api endpoint so `base_url` is optional. Tested locally and works.

davidsbond commented:
Hey @kelcya, thanks for addressing the comments.

For the documentation, I should have been clearer. The file here: https://github.com/davidsbond/terraform-provider-tailscale/blob/master/docs/index.md is what ends up on the terraform registry. Could you please add some information on the new provider configuration values there too?

Thanks :)

kelcya commented:
> Hey @kelcya, thanks for addressing the comments.
> 
> For the documentation, I should have been clearer. The file here: https://github.com/davidsbond/terraform-provider-tailscale/blob/master/docs/index.md is what ends up on the terraform registry. Could you please add some information on the new provider configuration values there too?
> 
> Thanks :)

Good to know. Updated.

davidsbond commented on tailscale/provider.go:
```diff
@@ -31,6 +31,11 @@ func Provider(options ...ProviderOption) *schema.Provider {
+			"base_url": {
+				Type:        schema.TypeString,
+				DefaultFunc: schema.EnvDefaultFunc("TAILSCALE_BASE_URL", "https://api.tailscale.com"),
+				Required:    true,
```
```suggestion
```

I don't know exactly what the behaviour will be if required is set to true and there's a default, so I don't think this should be a required field.

kelcya replied:
I have tried setting it to `false` in the past and removing it just now. Both yield the same result.

Either will prompt for the `base_url` if you do not specify in the provider block.

```
kyang@kyang terraform % terraform plan
provider.tailscale.base_url
  Enter a value: https://api.tailscale.com

╷
│ Error: InternalValidate
│
│   with provider["kyang.local/testing/tailscale"],
│   on <input-prompt> line 1:
│   (source code not available)
│
│ Internal validation of the provider failed! This is always a bug
│ with the provider itself, and not a user issue. Please report
│ this bug:
│
│ 1 error occurred:
│ 	* base_url: One of optional, required, or computed must be set
│
│
╵
```

Only setting it to require to true seems to work the way intended. If you do not specify the `base_url`, it will use the default.

davidsbond replied:
There should also be an `Optional` field you can set to true I think. I think you have to be explicit if a field is optional or required. A value of `false` in the `Required` field on its own does not indicate to the SDK that it is optional

davidsbond replied:
https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.14.0/helper/schema/schema.go#L44

Here's that type in particular. The `Optional` field should be settable here

kelcya replied:
Looks good. Tested locally.

davidsbond reviewed (commented):
Thanks, one last thing to change and I think this can be merfed

davidsbond commented:
Perfect. Thanks for addressing all the comments @kelcya I'll just wait for the tests to run then merge. If that's all good I'll publish a new release. A new release can take ~20 minutes to show up in the terraform registry once the release process is done.

kelcya commented:
> Perfect. Thanks for addressing all the comments @kelcya I'll just wait for the tests to run then merge. If that's all good I'll publish a new release. A new release can take ~20 minutes to show up in the terraform registry once the release process is done.

My pleasure. First time contributing to a public github repo. Thank you for writing this provider. I was using it to hack around Tailscale's lack of SCIM with our IDP but they recently rolled out a SCIM beta feature. Either way, it is useful. 👍

davidsbond commented:
Congratulations on your first contribution! That's great to hear :)
